vertx-pg-client/src/test/java/io/vertx/pgclient/TracingTest.java

```diff
@@ -0,0 +1,185 @@
+import io.vertx.sqlclient.impl.tracing.QueryRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+public class TracingTest extends PgTestBase {
+
+  Vertx vertx;
+  VertxTracer tracer;
+  PgPool pool;
+
+  @Before
+  public void setup() throws Exception {
+    super.setup();
+    vertx = Vertx.vertx(new VertxOptions().setTracingOptions(
+      new TracingOptions().setEnabled(true).setFactory(tracingOptions -> new VertxTracer() {
+        @Override
+        public Object sendRequest(Context context, Object request, String operation, BiConsumer headers, TagExtractor tagExtractor) {
+          return tracer.sendRequest(context, request, operation, headers, tagExtractor);
+        }
+
+        @Override
+        public void receiveResponse(Context context, Object response, Object payload, Throwable failure, TagExtractor tagExtractor) {
+          tracer.receiveResponse(context, response, payload, failure, tagExtractor);
+        }
+      }))
+    );
+    pool = PgPool.pool(vertx, options, new PoolOptions());
+  }
+
+  @After
+  public void teardown(TestContext ctx) {
+    vertx.close(ctx.asyncAssertSuccess());
+  }
+
+  @Test
+  public void testTraceSimpleQuery(TestContext ctx) {
+    String sql = "SELECT * FROM Fortune WHERE id=1";
+    testTraceQuery(ctx, sql, Collections.emptyList(), conn -> conn.query(sql).execute());
+  }
+
+  @Test
+  public void testTracePreparedQuery(TestContext ctx) {
+    String sql = "SELECT * FROM Fortune WHERE id=$1";
+    Tuple tuple = Tuple.of(1);
+    testTraceQuery(ctx, sql, Collections.singletonList(tuple), conn -> conn.preparedQuery(sql).execute(tuple));
+  }
+
+  @Test
+  public void testTraceBatchQuery(TestContext ctx) {
+    String sql = "SELECT * FROM Fortune WHERE id=$1";
+    List<Tuple> tuples = Arrays.asList(Tuple.of(1), Tuple.of(2));
+    testTraceQuery(ctx, sql, tuples, conn -> conn.preparedQuery(sql).executeBatch(tuples));
+  }
+
+  public void testTraceQuery(TestContext ctx, String expectedSql, List<Tuple> expectedTuples, Function<SqlClient, Future<?>> fn) {
+    AtomicBoolean called = new AtomicBoolean();
+    AtomicReference<Context> requestContext = new AtomicReference<>();
+    AtomicReference<Context> responseContext = new AtomicReference<>();
+    Object expectedPayload = new Object();
+    tracer = new VertxTracer<Object, Object>() {
+      @Override
+      public <R> Object sendRequest(Context context, R request, String operation, BiConsumer<String, String> headers, TagExtractor<R> tagExtractor) {
+        QueryRequest query = (QueryRequest) request;
+        ctx.assertEquals(query.sql(), expectedSql);
```

Comment:
in the assertions order of the expected value and actual value is reversed which reports wrong error messages when tests fail.

Reply:
right :-)